### PR TITLE
feat: :sparkles: add dynamic postal code support for shipping calculation

### PR DESCRIPTION
**Pull request by kennedysmartins**
#### Summary

This PR introduces the **dynamic postal code** feature, which allows shipping calculations to vary based on the `originCEP` defined in each product variant's metadata. The goal of this feature is to provide more accurate and granular shipping rates by grouping items according to their origin postal codes, rather than using a single static value for all items in the cart.

#### Key Features

- **Dynamic Postal Code per Item**: Each item in the cart can now have a distinct `originCEP`, specified in the product variant's `metadata`. If the variant contains an `originCEP`, it will override the default `MELHOR_ENVIO_POSTAL_CODE` defined in the environment variables.
  
- **Fallback Mechanism**: In cases where an `originCEP` is not provided in the variant metadata, the system will continue to use the default postal code from the environment variable `MELHOR_ENVIO_POSTAL_CODE`.

- **Grouped Shipping Calculation**: Items are grouped based on their `originCEP` for shipping rate calculations, ensuring that items shipped from different locations get accurate rates for their specific shipping origins.

#### Why This Approach?

We chose to implement this feature because many stores have inventory spread across multiple locations. A static origin postal code doesn’t account for this, leading to inaccurate shipping rates or potentially higher costs for customers. By allowing items to be grouped by their postal codes, we enable a more flexible and accurate shipping experience.

#### Discussion Point: `originCEP` Placement

Currently, the `originCEP` is stored in the **variant's metadata**. This decision was made to maintain flexibility on a per-item basis and ensure backward compatibility with existing setups. However, I'm open to suggestions on whether we should store this information in a more generic location within the system or if there are other best practices for managing shipping origins across different systems. 

Do you think the `metadata` is the best place for the `originCEP`, or should we consider defining a more universal method for handling multiple shipping origins?

### CODE REVIEW

File: src/services/melhor-envio-fulfillment.ts [hunk 1]
```diff
@@ -108,51 +108,81 @@ class MelhorEnvioFulfillmentService extends AbstractFulfillmentService {
...
+        return groups;
       },
-      services: optionData.external_id.toString(),
-    };
+      {}
+    );
+
+    let totalShippingCost = 0;
+
+    for (const originCEP in groupedProducts) {
```
Comment (wallycoding) [inline src/services/melhor-envio-fulfillment.ts]:
```suggestion
    for (const postalCodeOrigin in groupedProducts) {
```

File: src/services/melhor-envio-fulfillment.ts [hunk 2]
```diff
@@ -108,51 +108,81 @@ class MelhorEnvioFulfillmentService extends AbstractFulfillmentService {
...
-        postal_code: toPostaCode,
-      },
-      products: products,
-      options: {
-        receipt: false,
-        own_hand: false,
+    const groupedProducts: Record<string, any[]> = cart.items.reduce(
+      (groups, item) => {
+        const productDetails = getProductDetails(item);
+        const originCEP = productDetails.originCEP;
```
Comment (wallycoding) [inline src/services/melhor-envio-fulfillment.ts]:
```suggestion
        const postalCodeOrigin = productDetails.originCEP;
```

**Review by wallycoding — CHANGES_REQUESTED** [review]

**Comment by kennedysmartins** [comment]
Hi everyone,

After further consideration, I've decided to close this PR. I discovered a multi-warehouse module that may provide a more effective solution for handling shipping calculations, rather than relying on metadata for the origin postal code.

I will focus on implementing that approach, and once it’s ready, we can revisit the shipping calculation functionality.

Thank you for your understanding!

**Comment by marcosgomesneto** [comment]
@kennedysmartins  I liked the idea. See if you'd rather wait—do you think this multi-warehouse module does what you need? I'll wait then before proceeding with the pull request

**Comment by kennedysmartins** [comment]
@marcosgomesneto  Actually, I plan to test the multi-warehouse module to ensure we're using the correct field instead of relying on metadata for the origin postal code. My goal is to avoid potential complications that could arise from using metadata, ensuring that the shipping calculations are more accurate and easier to maintain in the long term. Once I've confirmed that the multi-warehouse module works as expected, we can proceed with the changes and make sure everything aligns with the intended functionality. I'll keep you updated on the progress.

**Pull request closed without merging**